desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java

```diff
@@ -88,7 +147,249 @@ public void onItemSelected(Datastore item) {
         Datastore currentValue = (Datastore) componentBuilder.getConfiguredProperty(propertyDescriptor);
         setValue(currentValue);
 
-        add(_comboBox);
+        _createDatastoreButton = WidgetFactory.createSmallPopupButton("",
+                IconUtils.ACTION_CREATE_TABLE);
+        _createDatastoreButton.setToolTipText("Create datastore");
+        JPopupMenu createDatastoreMenu = _createDatastoreButton.getMenu();
+        populateCreateDatastoreMenu(createDatastoreMenu);
+
+        _panelAroundButton = DCPanel.around(_createDatastoreButton);
+        _panelAroundButton.setBorder(WidgetUtils.BORDER_EMPTY);
+        _panelAroundButton.setVisible(true);
+
+        final DCPanel panel = new DCPanel();
+        panel.setLayout(new BorderLayout());
+        panel.add(_comboBox, BorderLayout.CENTER);
+        panel.add(_panelAroundButton, BorderLayout.EAST);
+
+        add(panel);
+    }
+    
+    public void setOnlyUpdatableDatastores(boolean onlyUpdatableDatastores) {
+        _onlyUpdatableDatastores = onlyUpdatableDatastores;
+        final JPopupMenu menu = _createDatastoreButton.getMenu();
+        menu.removeAll();
+        populateCreateDatastoreMenu(menu);
+    }
+
+    private void populateCreateDatastoreMenu(JPopupMenu createDatastoreMenu) {
+        List<String> alreadyAddedDatabases = new ArrayList<>();
+
+        final JMenuItem csvMenuItem = WidgetFactory.createMenuItem("CSV file", IconUtils.CSV_IMAGEPATH);
+        csvMenuItem.addActionListener(createActionListener(CsvDatastore.class, CsvDatastoreDialog.class));
+        createDatastoreMenu.add(csvMenuItem);
+        alreadyAddedDatabases.add("CSV file");
```

Comment:
Yeah, these are pretty nasty... Let's keep it as-is for now, but I think we have a little technical debt here.